server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java

```diff
@@ -109,9 +98,9 @@ protected void processInput(DesiredBalanceInput desiredBalanceInput) {
                 if (isFresh) {
                     if (shouldReroute) {
                         rerouteServiceSupplier.get().reroute("desired balance changed", Priority.NORMAL, ActionListener.noop());
-                    } else if (allocating == false) {
+                    } else {
                         logger.trace("Executing listeners up to [{}] as desired balance did not require reroute", lastConvergedIndex);
-                        executeListeners(lastConvergedIndex, desiredBalanceInput.routingAllocation().nodes().isLocalNodeElectedMaster());
+                        queue.complete(lastConvergedIndex);
```

Comment:
This doesn't seem quite right to me - AFAICT we could still be waiting on a previously-submitted reroute here. In more detail:

- Computation 42 converges with changes (`isFresh == true && shouldReroute == true`) and submits a reroute task.
- Computation 43 converges with no changes (`isFresh == true && shouldReroute == false`) and completes all listeners ≤43, including listener 42 even though its reroute hasn't completed.

Reply:
Queue#complete will complete the listeners immediately only if the `paused = false`. 

Queue is paused in the beginning of allocation cycle (triggered by the reroute) and resumed back after reconciler is complete. resume will execute listeners (if any) that are marked as complete during the time the queue was paused.

I believe this should cover above scenario.

Reply:
Oh, do you mean computation 43 is not caused by a reroute but a separate allocation? Yeah, I do not think this is handled by old nor new implementation yet

Reply:
I think that it might pick up a previous version of desired balance and reconciliate against it. Then this means listener is potentially executed too early.